src/compressed_tensors/compressors/pack_quantized.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import math
+from typing import Dict, Generator, Tuple
+
+import numpy as np
+import torch
+from compressed_tensors.compressors import Compressor
+from compressed_tensors.config import CompressionFormat
+from compressed_tensors.quantization import QuantizationArgs
+from compressed_tensors.quantization.lifecycle.forward import dequantize, quantize
+from compressed_tensors.quantization.utils import get_torch_bit_depth
+from compressed_tensors.utils import get_nested_weight_mappings, merge_names
+from safetensors import safe_open
+from torch import Tensor
+from tqdm import tqdm
+
+
+__all__ = ["PackedQuantizationCompressor", "pack_4bit_ints", "unpack_4bit_ints"]
+
+_LOGGER: logging.Logger = logging.getLogger(__name__)
+
+
+@Compressor.register(name=CompressionFormat.pack_quantized.value)
+class PackedQuantizationCompressor(Compressor):
+    """
+    Compresses a quantized model by packing every eight 4-bit weights into an int32
```

Comment:
So the benefit here comes from the fact that pytorch/numpy doesn't actually have an int4 type. When we quantize to 4 bits, we're still storing them in 8 bit integers and just wasting the top 4 bits. By packing the 4 bits into int32 we get a 2x compression ratio (roughly, this doesn't account for padding) compared to the "unpacked" int8 compression because we're eliminating the 4 unused bits in each int8

Reply:
And to clarify, we aren't packing away the 0s here like we do for the sparse bitmask compression

Reply:
In the future we could support packed quantization + sparse bitmask for extra compression but that would be more work because we have to account for the 0 points of the quantized data

Reply:
Are you aware of this: https://github.com/pytorch/ao/blob/aa9463960ad22c056b5e7e7472ad456e23756a9f/torchao/dtypes/uint4.py ?

Reply:
ah!! did not know about this. But either way to ask on the vLLM side was to pack the int4s into int32. Will see if we can use this

Reply:
There was some pressure on pytorch folks to introduce int4, i think this is as far as they went ;)